rfc/rfc-79/rfc-79.md

```diff
@@ -0,0 +1,104 @@
+
+### Concurrent writers can execute table service plan 
+
+When a writer executes a compact, cluster, or logcompaction plan, it will first rollback any existing inflight attempt, and (depending on the type and configuration of the table service) optionally re-execute it. This can lead to dataset corruption if one writer is rolling back the instant while another is still executing it. See https://issues.apache.org/jira/browse/HUDI-7503 for an example.
+Independent/outside of HUDI, a user may have an orchestration setup of concurrent writers where sometimes multiple writers can execute the sample plan simultaneously, due to a transient failure (at the orchestration leve) or misconfiguration. While it can be argued that HUDI does not to provide any guarantees of correctness if concurrent writers execute the same plan, updating HUDI to guard against this scenario would reduce the operational overhead and complexity of deploying a large-scale HUDI writer orchestration.
+
+### Removable-plan table service can be rolled back by clean before it can be executed
+
+After a writer schedules a remove-plan table service (such as logcompaction), another writer performing clean can rollback the plan before it has a chance to be executed, deleting the plan from the timeline. This will prevent the table service from ever being executed, and if the table service execution to be async (being deferred for execution later by another writer) then the chance of this happening increases.
+
+## Implementation
+In order to resolve these limitations with compact, logcompact, and cluster, a new configuration value `table_service_rollback_delay` will be introduced. The value will indicate the number of minutes that must elapse before a clean attempt is allowed to start a rollback of any removable-plan table service (logcompaction or cluster if they are configured as such). In addition, changes related to heartbeating and table lock transactions will be made to 
+the compact/logcompact/cluster APIs that execute a table service plan
+clean execution 
+ Combining all these changes, the logic for cluster/compact/logcompact table service execution and clean will now behave the following ways:
+
+### Changes to compaction/cluster plan exection
+The compact/logcompact/cluster API that executes a plan will now have the following steps performed before calling the current execution logic
+1. Get the table service plan P (as usual)
+2. Start a transaction
+3. Check if the instant time P has an active heartbeat, if so abort transaction and throw an exception
+4. If P is a  removable-plan, check if (a) P is an inflight instant or (b) there is a rollback plan on active timeline that targets P. If either is the case, then hrow an exception and abort the transaction. This is since this instant cannot be executed again but must be rolled back.
```

Comment:
Still don't get why the plan needs to be removable?

Reply:
I kind of think the hearbeat check should be done for any table sercive plan (no matter whether it is removable or not), and after the check, if the plan execution is allowed, we should put a heartbear immediately in case to unlock the other table service execution jobs.